File: app/src/main/java/co/cubeflow/parakeet/ui/AssignmentCreationFragment.java
package co.cubeflow.parakeet.ui;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TextView;

import org.joda.time.DateTime;

import java.util.Calendar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import co.cubeflow.parakeet.R;
import co.cubeflow.parakeet.model.Assignment;
import co.cubeflow.parakeet.ui.viewmodel.AssignmentCreationViewModel;

/**
 * A {@link Fragment} that provides options for creating a
 * {@link Assignment}.
 *
 * @version 1.0.0
 * @since 1.0.0
 */
public class AssignmentCreationFragment extends Fragment {

    private TextView mDueDateView;

    private TextView mGroupMembersView;

    private EditText mPointsView;

    private EditText mNotesView;

    /**
     * Required public empty constructor
     */
    public AssignmentCreationFragment() {

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_assignment_creation, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        mDueDateView = view.findViewById(R.id.assignment_due_date_label);
        mGroupMembersView = view.findViewById(R.id.group_members_content);
        mPointsView = view.findViewById(R.id.assignment_points_content);
        mNotesView = view.findViewById(R.id.assignment_notes_content);

        mDueDateView.setOnClickListener(v -> {
            Calendar calendar = Calendar.getInstance();
            int year = calendar.get(Calendar.YEAR);
            int month = calendar.get(Calendar.MONTH);
            int day = calendar.get(Calendar.DAY_OF_MONTH);
            //noinspection ConstantConditions Context won't be null.
            new DatePickerDialog(getContext(), (DatePicker picker, int y, int m, int d) -> {
                long dueDate = new DateTime(y, m + 1, d, 0, 0).getMillis();
                getViewModel().getDueDate().setValue(dueDate);
            }, year, month, day).show();
        });
    }

    @Override
    public void onStart() {
        super.onStart();
        getViewModel().getDueDate().observe(this, dueDate -> {
            DateTime dateTime = new DateTime(dueDate);
            mDueDateView.setText(dateTime.toString()); // TODO: 11/23/2017 Localize date string
        });

        getViewModel().getNotes().observe(this, notes -> {
            mNotesView.setText(notes);
        });

        mNotesView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // no-op
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
//                getViewModel().getNotes().setValue(String.valueOf(charSequence));
            }

            @Override
            public void afterTextChanged(Editable editable) {
                // no-op
            }
        });

        mPointsView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // no-op
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                getViewModel().getTotalPoints().setValue(Integer.valueOf(String.valueOf(charSequence)));
            }

            @Override
            public void afterTextChanged(Editable editable) {
                // no-op
            }
        });
    }

    @NonNull
    private AssignmentCreationViewModel getViewModel() {
        return ViewModelProviders.of(this).get(AssignmentCreationViewModel.class);
    }
}

File: app/src/main/java/co/cubeflow/parakeet/ui/AssignmentCreationFragment.kt
package co.cubeflow.parakeet.ui

import android.app.DatePickerDialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import co.cubeflow.parakeet.R
import co.cubeflow.parakeet.model.Assignment
import co.cubeflow.parakeet.ui.viewmodel.AssignmentCreationViewModel
import kotlinx.android.synthetic.main.fragment_assignment_creation.*
import org.joda.time.DateTime
import java.util.*

/**
 * A [Fragment] that provides options for creating a
 * [Assignment].
 *
 * @version 1.0.0
 * @since 1.0.0
 */
class AssignmentCreationFragment : Fragment() {

    private lateinit var viewModel: AssignmentCreationViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View =
            inflater.inflate(R.layout.fragment_assignment_creation, container, false)


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders.of(this).get(AssignmentCreationViewModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        assignment_due_date_label.setOnClickListener {
            val calendar = Calendar.getInstance()
            val year = calendar.get(Calendar.YEAR)
            val month = calendar.get(Calendar.MONTH)
            val day = calendar.get(Calendar.DAY_OF_MONTH)

            DatePickerDialog(context, { _, y, m, d ->
                val dueDate = DateTime(y, m + 1, d, 0, 0).millis
                viewModel.dueDate.setValue(dueDate)
            }, year, month, day).show()
        }
    }

    override fun onStart() {
        super.onStart()
        viewModel.dueDate.observe(this, Observer{ dueDate ->
            val dateTime = DateTime(dueDate)
            assignment_due_date_label!!.text = dateTime.toString() // TODO: 11/23/2017 Localize date string
        })

        viewModel.notes.observe(this, Observer {
            notes -> assignment_notes_content.setText(notes)
        })

        assignment_notes_content.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // no-op
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                //                getViewModel().getNotes().setValue(String.valueOf(charSequence));
            }

            override fun afterTextChanged(editable: Editable) {
                // no-op
            }
        })

        assignment_points_content.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // no-op
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                viewModel.totalPoints.setValue(Integer.valueOf(charSequence.toString()))
            }

            override fun afterTextChanged(editable: Editable) {
                // no-op
            }
        })
    }
}
